test/e2e/knit_irqwatch_test.go

```diff
@@ -0,0 +1,227 @@
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+	"github.com/openshift-kni/debug-tools/pkg/irqs"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+)
+
+var _ = g.Describe("knit IRQ watch tests", func() {
+
+	var fixtureName = "xeon-multinuma-00"
+	var snapshotRoot string
+
+	type cmdOutput struct {
+		out []byte
+		err error
+	}
+
+	g.Context("Without isolated CPUs", func() {
+		g.It("Produces the expected delta between consecutive irq samples", func() {
+			cmdline := []string{
+				filepath.Join(binariesPath, "knit"),
+				"-P", filepath.Join(snapshotRoot, "proc"),
+				"-S", filepath.Join(snapshotRoot, "sys"),
+				"-W", "3s",
+				"-T", "1",
+				"-J",
+				"irqwatch",
+			}
+			fmt.Fprintf(g.GinkgoWriter, "running: %v\n", cmdline)
+
+			cmd := exec.Command(cmdline[0], cmdline[1:]...)
+			cmd.Stderr = g.GinkgoWriter
+
+			c := make(chan cmdOutput)
+			go func(c chan cmdOutput) {
+				out, err := cmd.Output()
+				c <- cmdOutput{out: out,
+					err: err,
+				}
+			}(c)
+
+			// while the irqwatch is running change the interrupts file and return the changes
+			// this is very unlikley to suffer from race condition, since the irqwatch will wait 3 full seconds between each sample,
+			// so modifyInterruptsFile function will finish it's job already
+			delta, err := modifyInterruptsFile(filepath.Join(snapshotRoot, "proc"))
+			o.Expect(err).ToNot(o.HaveOccurred())
+
+			res := <-c
+			o.Expect(res.err).ToNot(o.HaveOccurred())
+
+			counters, err := extractCountersFromIRQWatchOutput(res.out)
+			o.Expect(res.err).ToNot(o.HaveOccurred())
+
+			ok := areCountersEqual(delta, counters)
+			o.Expect(ok).To(o.BeTrue())
+		})
+	})
+
+	g.BeforeEach(func() {
+		snapshotRoot = snapshotBeforeEach(fixtureName, "sysinfo.tgz")
+	})
+
+	g.AfterEach(func() {
+		snapshotAfterEach(snapshotRoot)
+	})
+})
+
+func modifyInterruptsFile(procFs string) (irqs.Stats, error) {
+	mockLog := log.New(ioutil.Discard, "", 0)
+
+	ih := irqs.New(mockLog, procFs)
+	stats, err := ih.ReadStats()
+	if err != nil {
+		return nil, err
+	}
+	initialStats := stats.Clone()
+	delta := make(irqs.Stats)
+	rand.Seed(time.Now().UnixNano())
+
+	// 50 random changes are good enough
+	const numOfModifications int = 50
+	// just a number that is big enough to make nice distribution of interrupts on random queues
+	const maxInterrupts = 1000
+
+	for i := 0; i < numOfModifications; i++ {
+		numOfCpus := len(stats)
+
+		// select random cpuid to change its counters
+		cpuid := int(rand.Intn(numOfCpus))
+
+		// select random IRQ in the seleced cpuid
+		k := selRandKey(stats[cpuid])
+
+		stats[cpuid][k] += uint64(rand.Intn(maxInterrupts))
+		delta[cpuid] = make(irqs.Counter)
+		delta[cpuid][k] = stats[cpuid][k] - initialStats[cpuid][k]
+	}
+
+	// add the random new stat values to the interruts file
+	return delta, reWriteInterrupts(procFs, stats)
+}
+
+func reWriteInterrupts(procFs string, stats irqs.Stats) error {
+	tmpInterrruptsFile := filepath.Join(procFs, "interrupts_temp")
```

Comment:
I'd use the utilities in `io/ioutil`

Reply:
not sure I follow, what alternative has to ```io/ioutil``` to offer?

Reply:
https://golang.org/pkg/io/ioutil/#TempFile + `os.Rename` once done. This is pretty much the standard pattern